### PR TITLE
ElasticTransform and RandomGaussianBlur transforms

### LINKED ISSUE #3: Elastic deformations
Elastic deformations, as described in the U-Net paper, are a much-needed addition to the coordinate-based augmentation methods in elektronn3.
This screenshot from the video at https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/ shows the general idea:
elastic

A 2D prototype demo for this augmentation method can be found at https://gist.github.com/mdraw/95e7e204e17eaa08c07735a008318acf (needs to be run locally with `jupyter notebook` to work).
We need to add 3D support to this and integrate it into the source coordinate warping pipeline.

### PR DESCRIPTION
**Pull request by mdraw**
Adding two new augmentation methods implemented by @mahsabh

Resolves #3.

### CODE REVIEW

File: elektronn3/data/transforms/transforms.py [hunk 1]
```diff
@@ -436,6 +496,76 @@ def __call__(
...
+            return inp, target
+        channels = range(inp.shape[0]) if self.channels is None else self.channels
+        deformed_img = np.empty_like(inp)
+        for c in channels:
+            shape = inp[c].shape
+            if inp[c].ndim ==3 :
+                dx = gaussian_filter((self.rng.rand(*shape) * 2 - 1), self.sigma, mode="constant", cval=0) * self.alpha
+                dy = gaussian_filter((self.rng.rand(*shape) * 2 - 1), self.sigma, mode="constant", cval=0) * self.alpha
+                dz = gaussian_filter((self.rng.rand(*shape) * 2 - 1), self.sigma, mode="constant", cval=0) * self.alpha
+                x, y, z = np.meshgrid(np.arange(shape[0]), np.arange(shape[1]), np.arange(shape[2]), indexing='ij')
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]: `x, y, z` and `x, y` below are misleading because we always expect image data in the dimension order (C, D, H, W) for 3D or (C, H, W) for 2D images. So if you need to explicitly name the dimensions, this line should be `z, y, x = ...`. Same for `dx` etc.

File: elektronn3/data/transforms/transforms.py [hunk 2]
```diff
@@ -436,6 +496,76 @@ def __call__(
...
+                dx = gaussian_filter((self.rng.rand(*shape) * 2 - 1), self.sigma, mode="constant", cval=0) * self.alpha
+                dy = gaussian_filter((self.rng.rand(*shape)* 2 - 1), self.sigma, mode="constant", cval=0) * self.alpha
+                x, y = np.meshgrid(np.arange(shape[0]), np.arange(shape[1]))
+                indices = np.reshape(y + dy, (-1, 1)), np.reshape(x + dx, (-1, 1))
+
+            else:
+                raise ValueError("Image dimension not understood!")
+
+            deformed_img[c] = map_coordinates(inp[c], indices, order=1).reshape(shape)
+
```
[the remark this answers is not in the document]
Comment (mahsabh) [reply elektronn3/data/transforms/transforms.py]: I guess I am not clear yet what we mean by 'target'. Where is it explained in the documentation ? 
or is it a different concept for each function ?

File: elektronn3/data/transforms/transforms.py (hunk 1 above)
Comment (mahsabh) [reply elektronn3/data/transforms/transforms.py]: hmm... I remember there was some confusion here. Specially with the reshape function. I tried different variants until it worked correctly! But I will change it and see if I can get it work in that order.

File: elektronn3/data/transforms/transforms.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdraw) [reply elektronn3/data/transforms/transforms.py]: The target is the tensor that contains the ground truth labels for the given input tensor, i.e. the desired output of a neural network for this input. We always calculate the loss as a function of output and target tensor, see https://github.com/ELEKTRONN/elektronn3/blob/master/elektronn3/training/trainer.py#L447 and
https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss

[the remark this answers is not in the document]
Comment (mahsabh) [reply elektronn3/data/transforms/transforms.py]: Do you want to have the the different modes for map_coordinates as an argument?

[the remark this answers is not in the document]
Comment (mdraw) [reply elektronn3/data/transforms/transforms.py]: We're unlikely to keep using SciPy's `map_coordinates` in the long term because it's prohibitively slow for 3D images, so we shouldn't add any special options for it (we'll try to integrate this with the numba-based warping code in https://github.com/ELEKTRONN/elektronn3/blob/master/elektronn3/data/coord_transforms.py once this PR is merged).

You can keep it hardcoded to `order=1` for images and use a flag (see https://github.com/ELEKTRONN/elektronn3/pull/26#discussion_r259632435) to control if targets should be mapped with `order=0` ("discrete") or `order=1` ("continuous").

File: elektronn3/data/transforms/transforms.py [hunk 3]
```diff
@@ -436,6 +496,102 @@ def __call__(
...
+            if target is None:
+                return deformed_img
+            else:
+                assert inp.shape == target.shape
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]: Assertions are fine for self-explanatory stuff, but if `inp.shape != target.shape`, a `NotImplementedError` should be raised instead that says that differently-shaped targets are not (yet) supported in `ElasticTransform`.

File: elektronn3/data/transforms/transforms.py [hunk 4]
```diff
@@ -436,6 +496,102 @@ def __call__(
...
+            channels: If ``channels`` is ``None``, the change is applied to
+                all channels of the input tensor.
+                If ``channels`` is a ``Sequence[int]``, change is only applied
+                to the specified channels.
+
+            prob: probability (between 0 and 1) with which to perform this
+                augmentation. The input is returned unmodified with a probability
+                of ``1 - prob``
+
+            target_discrete: bool
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]: To stay consistent, please make this a list of bools instead, where the entries refer to the different channels of the target tensor. See https://github.com/ELEKTRONN/elektronn3/pull/26/files#r259632435

File: elektronn3/data/transforms/transforms.py [hunk 5]
```diff
@@ -436,6 +496,102 @@ def __call__(
...
+
+                This information is used to decide what kind of interpolation should
+                be used for reading target data:
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]: There has to be an empty line between this line and the next one because Sphinx (which renders the documentation from docstrings) for some reason requires an empty line before enumerations so it can render it correctly. See http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks

On the other hand all the empty lines between the different args can (and should) be removed so the docstring takes up less lines.

File: elektronn3/data/transforms/transforms.py [hunk 6]
```diff
@@ -436,6 +496,102 @@ def __call__(
...
+                y, x = np.meshgrid(np.arange(shape[0]), np.arange(shape[1]))
+                indices = np.reshape(x + dx, (-1, 1)), np.reshape(y + dy, (-1, 1))
+
+            else:
+                raise ValueError("Image dimension not understood!")
+
+            deformed_img[c] = map_coordinates(inp[c], indices, order=1).reshape(shape)
+
+            if target is None:
+                return deformed_img
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]: You can't return the `deformed_img` alone, because this would break the assumption that a tuple of input and target is returned (where the target can be `None`, but it still has to be returned because of the function signature).

Comment (mahsabh) [reply elektronn3/data/transforms/transforms.py]: Of course! Sorry! I must have missed it.

File: elektronn3/training/trainer.py [hunk 7]
```diff
@@ -32,6 +32,7 @@
 from elektronn3.inference import Predictor
 from elektronn3 import __file__ as arch_src
 
+
```
Comment (mdraw) [inline elektronn3/training/trainer.py]:
```suggestion
```

File: elektronn3/data/transforms/transforms.py [hunk 8]
```diff
@@ -436,6 +496,128 @@ def __call__(
...
+            if self.target_discrete_ix is None:
+                self.target_discrete_ix = [True for i in range(target_channels)]
+            else:
+                self.target_discrete_ix = [i in self.target_discrete_ix for i in range(target_channels)]
+
+            deformed_target = np.empty_like(target)
+            if target_c == True:
+                for tc in range(target_channels):
+                    target_order = 0 if self.target_discrete_ix[tc] is True else 1
+                    deformed_target[tc] = map_coordinates(target[tc], indices, order=target_order ).reshape(shape)
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]: I don't think you want to reshape the `deformed_target` to `shape` , because that's the input shape (which can be the same, but doesn't have to be).

File: elektronn3/data/transforms/transforms.py [hunk 9]
```diff
@@ -436,6 +496,128 @@ def __call__(
...
+                    target_c = False
+                    target_channels = 1
+            #if target is not None do the followings
+            if self.target_discrete_ix is None:
+                self.target_discrete_ix = [True for i in range(target_channels)]
+            else:
+                self.target_discrete_ix = [i in self.target_discrete_ix for i in range(target_channels)]
+
+            deformed_target = np.empty_like(target)
+            if target_c == True:
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]:
```suggestion
            if target_c:
```

File: elektronn3/data/transforms/transforms.py [hunk 10]
```diff
@@ -436,6 +496,128 @@ def __call__(
...
+            target_c = True  # True if the first dim of target is the number of channels
+            if target.ndim == 4:  # (C, D, H, W)
+                target_channels = target.shape[0]
+            elif target.ndim == 3:  # (C, H, W) or (D, H, W)
+                if inp.ndim == 3:  # (C, H, W)
+                    target_channels = target.shape[0]
+                elif inp.ndim == 4:  # (D, H, W)
+                    target_c = False
+                    target_channels = 1
+            #if target is not None do the followings
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]:
```suggestion
```
Should already be clear enough from looking at the code.

File: elektronn3/data/transforms/transforms.py (hunk 8 above)
Comment (mahsabh) [reply elektronn3/data/transforms/transforms.py]: you mean I should define its own indices ?

Comment (mdraw) [reply elektronn3/data/transforms/transforms.py]: No, the indices are shared with the target deformation, that's already correct. I'm only referring to the `.reshape(shape)` part at the end of the line.

File: elektronn3/data/transforms/transforms.py [hunk 11]
```diff
@@ -436,6 +496,131 @@ def __call__(
...
+        self.alpha = alpha
+        self.channels = channels
+        self.prob = prob
+        self.rng = np.random.RandomState() if rng is None else rng
+        self.target_discrete_ix = target_discrete_ix
+
+    def __call__(
+            self,
+            inp: np.ndarray,
+            target: Optional[np.ndarray] = None  # returned without modifications
```
Comment (mdraw) [inline elektronn3/data/transforms/transforms.py]:
```suggestion
            target: Optional[np.ndarray] = None
```
It's not returned without modification if it's not `None`

**Comment by mdraw** [comment]
LGTM, thanks! Manually merged in c6bf136d.
